File: apps/mcp-gateway/src/managed/runtime-config/validation.ts
import {
  McpProxyExposure, McpProxyToolClassification, McpProxyTransport, ProxyApprovalMode,
  type RuntimeConfiguration,
} from "@apex/contracts";
import { hash, httpsUrl, identifier, reference, requireValue, textBound, unique } from "./boundary.js";
import { hostName, validateNetwork } from "./network.js";
import { schemaJson } from "./schemas.js";

const UUID_V7 = /^[0-9a-f]{8}-[0-9a-f]{4}-7[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/;

/** Metadata parity with proxy/runtime_config and existing domain validation.
 * The catalog and approved output-profile registry are NOT in this wire message:
 * their authority must come from trusted provisioning, never a self-signed hash.
 * Full schema/output enforcement and admission remain later runtime stages.
 */
export function validateMetadata(config: RuntimeConfiguration): void {
  requireValue(config.schemaVersion === 1 && config.generation > 0n);
  requireValue(identifier(config.workspaceId, 256) && identifier(config.namespaceId, 256));
  requireValue(UUID_V7.test(config.proxyId) && UUID_V7.test(config.revisionId));
  requireValue(hash(config.configHash) && hash(config.runtimeManifestHash));
  const spec = config.spec;
  const ingress = spec?.ingress;
  const profile = spec?.runtimeProfile;
  const governance = spec?.governanceBinding;
  const auth = config.auth;
  const telemetry = config.telemetry;
  requireValue(spec && ingress && profile && governance && auth && telemetry);
  requireValue(ingress.transport === McpProxyTransport.STREAMABLE_HTTP && ingress.protocolRevision === "2025-11-25");
  requireValue([McpProxyExposure.PRIVATE, McpProxyExposure.EXTERNAL].includes(ingress.exposure));
  requireValue(ingress.inboundAuthenticationRequired && spec.cliProfiles.length === 0);
  requireValue(profile.rootless && profile.networkPolicy === "default-deny" && profile.filesystemPolicy === "read-only-rootfs");
  hostName(ingress.host);
  textBound(ingress.path);
  const resource = httpsUrl(config.resourceUrl);
  requireValue(resource.href === config.resourceUrl && resource.hostname === ingress.host);
  requireValue(resource.pathname === ingress.path && (!resource.port || resource.port === "443"));
  requireValue(auth.audience === config.resourceUrl);
  requireValue(ingress.allowedOrigins.length <= 32);
  unique(ingress.allowedOrigins);
  for (const origin of ingress.allowedOrigins) requireValue(httpsUrl(origin).pathname === "/");
  httpsUrl(auth.issuer);
  httpsUrl(auth.jwksUri);
  reference(auth.workloadIdentityRef, "identity://");
  requireValue(auth.requiredScopes.length > 0 && auth.requiredScopes.length <= 64);
  unique(auth.requiredScopes);
  requireValue(auth.requiredScopes.every(scope => identifier(scope)));

  requireValue(spec.upstreams.length > 0 && spec.upstreams.length <= 64);
  const upstreamIds = unique(spec.upstreams.map(upstream => upstream.upstreamId));
  const declaredSecrets = new Set<string>();
  function secret(value: string): void { reference(value, "secret://"); declaredSecrets.add(value); }
  for (const upstream of spec.upstreams) {
    requireValue(identifier(upstream.upstreamId) && upstream.transport === McpProxyTransport.STREAMABLE_HTTP);
    textBound(upstream.displayName);
    textBound(upstream.serverIdentity);
    requireValue(hash(upstream.toolCatalogHash) && upstream.secretRefs.length <= 32);
    const endpoint = httpsUrl(upstream.endpointOrCommandRef);
    requireValue(config.networkGrants.some(grant => endpoint.hostname === grant.host && Number(endpoint.port || "443") === grant.port));
    secret(upstream.credentialRef);
    unique(upstream.secretRefs);
    upstream.secretRefs.forEach(secret);
  }
  requireValue(spec.authBindings.length <= 32);
  unique(spec.authBindings.map(binding => binding.bindingId));
  for (const binding of spec.authBindings) {
    requireValue(identifier(binding.bindingId) && binding.scopes.length <= 64);
    textBound(binding.inboundSubject);
    secret(binding.outboundCredentialRef);
    unique(binding.scopes);
    requireValue(binding.scopes.every(scope => identifier(scope)));
  }
  requireValue(config.secretRefs.length <= 4096);
  const suppliedSecrets = unique(config.secretRefs);
  requireValue(suppliedSecrets.size === declaredSecrets.size && [...suppliedSecrets].every(ref => declaredSecrets.has(ref)));

  requireValue(spec.exposedTools.length > 0 && spec.exposedTools.length <= 256);
  unique(spec.exposedTools.map(tool => tool.alias));
  const toolKey = (upstream: string, tool: string): string => JSON.stringify([upstream, tool]);
  const exposed = new Set(spec.exposedTools.map(tool => toolKey(tool.upstreamId, tool.toolName)));
  for (const tool of spec.exposedTools) {
    requireValue(upstreamIds.has(tool.upstreamId) && identifier(tool.toolName) && identifier(tool.alias));
    requireValue([McpProxyToolClassification.READ, McpProxyToolClassification.BUSINESS_WRITE, McpProxyToolClassification.HIGH_IMPACT].includes(tool.classification));
    requireValue(tool.toolName !== "portfolio.read" || tool.classification === McpProxyToolClassification.READ);
  }
  requireValue(config.toolSchemas.length === exposed.size && config.toolSchemas.length <= 256);
  unique(config.toolSchemas.map(schema => toolKey(schema.upstreamId, schema.toolName)));
  for (const schema of config.toolSchemas) {
    requireValue(exposed.has(toolKey(schema.upstreamId, schema.toolName)));
    requireValue(identifier(schema.outputProfileId) && hash(schema.schemaHash));
    schemaJson(schema.inputSchemaJson);
    schemaJson(schema.outputSchemaJson);
  }

  requireValue(identifier(governance.policyId));
  const modes = new Map([
    ["none", ProxyApprovalMode.NONE], ["operator", ProxyApprovalMode.OPERATOR], ["dual-operator", ProxyApprovalMode.DUAL_OPERATOR],
  ]);
  requireValue(modes.has(governance.approvalMode) && modes.get(governance.approvalMode) === config.approvalMode);
  requireValue(["public", "internal", "confidential", "restricted"].includes(governance.dataClassification));
  limit(governance.rateLimit, "/m", 1000000n);
  limit(governance.concurrencyLimit, "", 1000000n);
  limit(governance.budget, "/d", 1000000n);
  limit(governance.retention, "d", 3650n);
  const cpu = profile.cpuLimit.endsWith("m") ? decimal(profile.cpuLimit.slice(0, -1)) : decimal(profile.cpuLimit) * 1000n;
  const unit = /^(\d+)(Ki|Mi|Gi)?$/.exec(profile.memoryLimit);
  requireValue(unit);
  const scale = new Map([["Ki", 1024n], ["Mi", 1048576n], ["Gi", 1073741824n]]);
  const memory = decimal(unit[1]) * (unit[2] ? scale.get(unit[2])! : 1n);
  requireValue(cpu >= 1n && cpu <= 4000n && BigInt(config.cpuMillis) === cpu);
  requireValue(memory >= 16777216n && memory <= 2147483648n && config.memoryBytes === memory);
  requireValue(config.pidLimit >= 16 && config.pidLimit <= 1024);
  image(config.imageRef, profile.imageDigest);
  validateNetwork(profile.egressDestinations, config.networkGrants);
  requireValue(telemetry.fullTraceSamplePerMillion >= 1 && telemetry.fullTraceSamplePerMillion <= 1000000);
  requireValue(telemetry.maxStages >= 1 && telemetry.maxStages <= 32);
  requireValue(telemetry.maxSummaryBytes >= 1 && telemetry.maxSummaryBytes <= 65536);
  requireValue(telemetry.maxSpans >= 1 && telemetry.maxSpans <= 128);
  requireValue(telemetry.maxAttributesPerSpan >= 1 && telemetry.maxAttributesPerSpan <= 64);
  requireValue(telemetry.maxExportQueueBytes >= 1n && telemetry.maxExportQueueBytes <= 8388608n);
}

function decimal(value: string): bigint {
  requireValue(value.length <= 512 && /^[0-9]+$/.test(value));
  const parsed = BigInt(value);
  requireValue(parsed <= (1n << 64n) - 1n);
  return parsed;
}

function limit(value: string, suffix: string, maximum: bigint): void {
  requireValue(value.endsWith(suffix));
  const digits = suffix ? value.slice(0, -suffix.length) : value;
  requireValue(/^[1-9][0-9]*$/.test(digits));
  requireValue(decimal(digits) <= maximum);
}

function image(value: string, digest: string): void {
  requireValue(/^sha256:[0-9a-f]{64}$/.test(digest) && value.length <= 512);
  const parts = value.split("@");
  requireValue(parts.length === 2 && parts[1] === digest);
  const slash = parts[0].indexOf("/");
  requireValue(slash > 0);
  const registry = parts[0].slice(0, slash);
  const repository = parts[0].slice(slash + 1);
  const url = httpsUrl(`https://${registry}/`);
  requireValue(registry.includes(".") && url.pathname === "/" && url.origin === `https://${registry}`);
  requireValue(repository.split("/").every(part => /^[a-z0-9][a-z0-9._-]*$/.test(part) && !part.includes("..")));
}

File: apps/control-plane-api/src/proxy/runtime_config/validation.rs
use std::collections::BTreeSet;

use super::{RuntimeDeploymentBindings, invalid, require};
use crate::proxy::{
    McpProxyRevision, ProxyError, ProxyTransport, SecretRef, validation::is_scope_identifier,
};

pub(super) fn deployment(
    revision: &McpProxyRevision,
    b: &RuntimeDeploymentBindings,
) -> Result<(), ProxyError> {
    require(
        is_scope_identifier(&b.scope.workspace_id) && is_scope_identifier(&b.scope.namespace_id),
    )?;
    require(b.generation > 0 && (16..=1024).contains(&b.pid_limit))?;
    let ingress = &revision.spec.ingress;
    let profile = &revision.spec.runtime_profile;
    // CLI/stdio need the later agent's approved executable and confinement
    // contract. Do not silently serialize an unsupported executable shape.
    require(
        ingress.transport == ProxyTransport::StreamableHttp
            && ingress.protocol_revision == "2025-11-25"
            && ingress.inbound_authentication_required
            && profile.rootless
            && profile.network_policy == "default-deny"
            && profile.filesystem_policy == "read-only-rootfs"
            && revision.spec.cli_profiles.is_empty(),
    )?;
    let resource = https_url(&b.resource_url)?;
    require(
        resource.as_str() == b.resource_url
            && resource.host_str() == Some(ingress.host.as_str())
            && resource.path() == ingress.path
            && resource.port_or_known_default() == Some(443)
            && b.auth.audience == b.resource_url,
    )?;
    for origin in &ingress.allowed_origins {
        let url = https_url(origin)?;
        require(url.path() == "/")?;
    }
    https_url(&b.auth.issuer)?;
    https_url(&b.auth.jwks_uri)?;
    reference(&b.auth.workload_identity_ref, "identity://")?;
    require(!b.auth.required_scopes.is_empty() && b.auth.required_scopes.len() <= 64)?;
    let mut scopes = BTreeSet::new();
    for scope in &b.auth.required_scopes {
        require(identifier(scope) && scopes.insert(scope))?;
    }
    for upstream in &revision.spec.upstreams {
        require(
            upstream.transport == ProxyTransport::StreamableHttp
                && upstream.tool_catalog_hash.as_deref().is_some_and(hex_hash),
        )?;
        let url = https_url(&upstream.endpoint_or_command_ref)?;
        require(b.network_grants.iter().any(|grant| {
            url.host_str() == Some(grant.host.as_str())
                && url.port_or_known_default().map(u32::from) == Some(grant.port)
        }))?;
    }
    let t = &b.telemetry;
    require(
        (1..=1_000_000).contains(&t.full_trace_sample_per_million)
            && (1..=32).contains(&t.max_stages)
            && (1..=65_536).contains(&t.max_summary_bytes)
            && (1..=128).contains(&t.max_spans)
            && (1..=64).contains(&t.max_attributes_per_span)
            && (1..=8_388_608).contains(&t.max_export_queue_bytes),
    )
}

pub(super) fn secret_references(
    revision: &McpProxyRevision,
    bindings: &[SecretRef],
) -> Result<Vec<String>, ProxyError> {
    let spec = &revision.spec;
    let declared: BTreeSet<_> = spec
        .upstreams
        .iter()
        .flat_map(|u| u.credential_ref.iter().chain(&u.secret_refs))
        .chain(
            spec.auth_bindings
                .iter()
                .flat_map(|b| b.outbound_credential_ref.iter()),
        )
        .chain(spec.cli_profiles.iter().flat_map(|p| &p.secret_refs))
        .map(SecretRef::as_str)
        .collect();
    require(bindings.len() <= 4096)?;
    let supplied: BTreeSet<_> = bindings.iter().map(SecretRef::as_str).collect();
    require(supplied.len() == bindings.len() && supplied == declared)?;
    for value in &declared {
        reference(value, "secret://")?;
    }
    Ok(declared.into_iter().map(str::to_owned).collect())
}

pub(super) fn image(
    revision: &McpProxyRevision,
    b: &RuntimeDeploymentBindings,
) -> Result<String, ProxyError> {
    require(!b.image_catalog.is_empty() && b.image_catalog.len() <= 256)?;
    let digest = &revision.spec.runtime_profile.image_digest;
    let image = b.image_catalog.get(digest).ok_or_else(invalid)?;
    require(image.len() <= 512)?;
    let (name, selected_digest) = image.split_once('@').ok_or_else(invalid)?;
    require(selected_digest == digest)?;
    let (registry, repository) = name.split_once('/').ok_or_else(invalid)?;
    let registry_url = https_url(&format!("https://{registry}/"))?;
    require(
        registry.contains('.')
            && registry_url.path() == "/"
            && registry_url.origin().ascii_serialization() == format!("https://{registry}"),
    )?;
    require(repository.split('/').all(|part| {
        !part.is_empty()
            && !part.contains("..")
            && part
                .as_bytes()
                .first()
                .is_some_and(u8::is_ascii_alphanumeric)
            && part.bytes().all(|c| {
                c.is_ascii_lowercase() || c.is_ascii_digit() || matches!(c, b'.' | b'_' | b'-')
            })
    }))?;
    Ok(image.clone())
}

pub(super) fn resource_units(cpu: &str, memory: &str) -> Result<(u32, u64), ProxyError> {
    let (cpu_digits, cpu_scale) = cpu.strip_suffix('m').map_or((cpu, 1000), |s| (s, 1));
    let cpu_millis = decimal(cpu_digits)?
        .checked_mul(cpu_scale)
        .ok_or_else(invalid)?;
    require((1..=4000).contains(&cpu_millis))?;
    let (memory_digits, scale) = if let Some(digits) = memory.strip_suffix("Ki") {
        (digits, 1024)
    } else if let Some(digits) = memory.strip_suffix("Mi") {
        (digits, 1_048_576)
    } else if let Some(digits) = memory.strip_suffix("Gi") {
        (digits, 1_073_741_824)
    } else {
        (memory, 1)
    };
    let memory_bytes = decimal(memory_digits)?
        .checked_mul(scale)
        .ok_or_else(invalid)?;
    require((16_777_216..=2_147_483_648).contains(&memory_bytes))?;
    Ok((
        u32::try_from(cpu_millis).map_err(|_| invalid())?,
        memory_bytes,
    ))
}

fn decimal(value: &str) -> Result<u64, ProxyError> {
    require(!value.is_empty() && value.bytes().all(|c| c.is_ascii_digit()))?;
    value.parse().map_err(|_| invalid())
}

pub(super) fn https_url(value: &str) -> Result<reqwest::Url, ProxyError> {
    require(
        !value.is_empty()
            && value.len() <= 512
            && !value
                .chars()
                .any(|c| c.is_whitespace() || c.is_control() || c == '\\'),
    )?;
    let url = reqwest::Url::parse(value).map_err(|_| invalid())?;
    require(
        url.scheme() == "https"
            && url.host_str().is_some()
            && url.username().is_empty()
            && url.password().is_none()
            && url.query().is_none()
            && url.fragment().is_none(),
    )?;
    Ok(url)
}

pub(super) fn reference(value: &str, prefix: &str) -> Result<(), ProxyError> {
    let tail = value.strip_prefix(prefix).ok_or_else(invalid)?;
    require(
        value.len() <= 512
            && !tail.is_empty()
            && tail
                .split('/')
                .all(|s| !s.is_empty() && s != "." && s != "..")
            && tail
                .as_bytes()
                .first()
                .is_some_and(u8::is_ascii_alphanumeric)
            && tail.bytes().all(|c| {
                c.is_ascii_alphanumeric() || matches!(c, b'.' | b'_' | b':' | b'/' | b'-')
            }),
    )
}

pub(super) fn identifier(value: &str) -> bool {
    value.len() <= 128 && is_scope_identifier(value)
}

pub(super) fn hex_hash(value: &str) -> bool {
    value.len() == 64
        && value
            .bytes()
            .all(|b| b.is_ascii_digit() || matches!(b, b'a'..=b'f'))
}
